rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+Kubernetes clusters from a single-host to enable the Kubernetes platform team
+to manage and monitor clusters across the organization.
+
+For users who do not leverage Machine ID, there are still some benefits of 
+undertaking this work. For example, using `tsh` to generate a Kubeconfig will
+be faster as it will not need to request the signing of a new certificate as 
+part of the process.
+
+Additionally, from an idealistic design perspective, the current approach
+compromises the intended use of certificates. Certificates are designed to
+encode attributes of an identity - information about which cluster you wish to
+connect to is not an attribute of identity!
+
+## Details
+
+### Today
+
+Today, we encode the intended target cluster within the user X.509 certificate
+within the `KubernetesCluster` attribute (OID 1.9999.1.3).  
+
+This attribute is primarily used by the Proxy and Kubernetes Agent to determine
+where to route requests intended for the Kubernetes cluster.
+
+When the KubernetesCluster attribute is requested by a client using the 
+`GenerateUserCerts` RPC, the certificate request is considered "Kubernetes"
+for analytics purposes. If we were to replace this attribute, we would need to 
+find an alternative way to determine if a certificate request is intended for
+Kubernetes.
+
+The target Kubernetes agent performs authorization checks based on the
+user's identity (e.g roleset, attributes) rather than relying on this attribute.
+
+### Proposal
+
+Rather than leverage the `KubernetesCluster` attribute within the X509
+certificate for routing requests to a specific Kubernetes cluster, we propose to
+pre-pend the request path with this information.
+
+For example, today, to make a request, the client may use the following URL:
+
+```shell
+GET https://example.teleport.sh/apis/apps/v1/namespaces/my-namespace/deployments
+```
+
+Following this proposal, the client would use the following URL:
+
+```shell
+GET https://example.teleport.sh/v1/teleport/base64-teleport-cluster-name/base64-k8s-cluster-name/apis/apps/v1/namespaces/my-namespace/deployments
+```
+
+To avoid potential problems with URL encoding/escaping, the name of the Teleport
+and Kubernetes cluster will be base64 encoded.
+
+Within the Kube Forwarder, new route handlers would be added for all HTTP
+methods for `/v1/teleport/:teleportCluster/:kubeCluster/*`. These route handlers
+would parse the `teleportCluster` and `kubeCluster` from the URL, strip off the
+prefix, and then pass the request through to the original handler - with the
+additional context of the desired Teleport and Kubernetes cluster. This would
+take higher precedence than attributes contained within the certificate.
+
+Teleport clients would no longer need to call `GenerateUserCerts` with a
+`KubernetesCluster` to produce a certificate explicitly for access to
+Kubernetes. When producing a kubeconfig, the client should use the user's 
+standard identity.
```

Comment:
Beyond just routing info (which this will be superseding) we also have a certificate usage restriction for Kubernetes, which the auth middleware currently requires for anything that touches the forwarder. Do you see the `usage:kube` restriction going away with this?

Reply:
The usage was only configured when using KubeCSR, which was deprecated several versions ago. 

Furthermore, if the identity does not have a usage specified, it will bypass any usage validation checks.

https://github.com/gravitational/teleport/blob/b71f017244e160e0b3eaef9c85c4dc230d2bd378/lib/auth/middleware.go#L664-L673